go/libraries/doltcore/env/actions/table.go

```diff
@@ -21,74 +21,127 @@ import (
 	"github.com/liquidata-inc/dolt/go/libraries/doltcore/env"
 )
 
-func CheckoutAllTables(ctx context.Context, dEnv *env.DoltEnv) error {
+func CheckoutAllTables(ctx context.Context, dEnv *env.DoltEnv) (unknown string, err error) {
 	roots, err := getRoots(ctx, dEnv, WorkingRoot, StagedRoot, HeadRoot)
 
 	if err != nil {
-		return err
+		return "", err
 	}
 
 	tbls, err := AllTables(ctx, roots[WorkingRoot], roots[StagedRoot], roots[HeadRoot])
 
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return checkoutTables(ctx, dEnv, roots, tbls)
+	docs := *env.AllValidDocDetails
+
+	return checkoutTablesAndDocs(ctx, dEnv, roots, tbls, docs)
 
 }
 
-func CheckoutTables(ctx context.Context, dEnv *env.DoltEnv, tbls []string) error {
+func CheckoutTablesAndDocs(ctx context.Context, dEnv *env.DoltEnv, tbls []string, docs []doltdb.DocDetails) (unknown string, err error) {
 	roots, err := getRoots(ctx, dEnv, WorkingRoot, StagedRoot, HeadRoot)
 
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return checkoutTables(ctx, dEnv, roots, tbls)
+	return checkoutTablesAndDocs(ctx, dEnv, roots, tbls, docs)
 }
 
-func checkoutTables(ctx context.Context, dEnv *env.DoltEnv, roots map[RootType]*doltdb.RootValue, tbls []string) error {
-	var unknown []string
+func checkoutTablesAndDocs(ctx context.Context, dEnv *env.DoltEnv, roots map[RootType]*doltdb.RootValue, tbls []string, docs []doltdb.DocDetails) (unknown string, err error) {
+	unknownTbls := []string{}
 
 	currRoot := roots[WorkingRoot]
 	staged := roots[StagedRoot]
 	head := roots[HeadRoot]
+
+	if len(docs) > 0 {
+		currRoot, staged, err = getUpdatedWorkingAndStagedWithDocs(ctx, dEnv, currRoot, staged, head, docs)
+		if err != nil {
+			return "", err
+		}
+	}
+
 	for _, tblName := range tbls {
+		if tblName == doltdb.DocTableName {
+			continue
+		}
 		tbl, ok, err := staged.GetTable(ctx, tblName)
 
 		if err != nil {
-			return err
+			return "", err
 		}
 
 		if !ok {
 			tbl, ok, err = head.GetTable(ctx, tblName)
 
 			if err != nil {
-				return err
+				return "", err
 			}
 
 			if !ok {
-				unknown = append(unknown, tblName)
+				unknownTbls = append(unknownTbls, tblName)
 				continue
 			}
 		}
 
 		currRoot, err = currRoot.PutTable(ctx, tblName, tbl)
 
 		if err != nil {
-			return err
+			return "", err
 		}
 	}
 
-	if len(unknown) > 0 {
+	if len(unknownTbls) > 0 {
+		for _, tbl := range unknownTbls {
```

Comment:
What's the purpose of this new block of code? Was this a bug in the previous implementation?

Reply:
RemoveTables silently ignores tables that aren't in the root, so this is a behavior change. Not saying this is wrong, just a change. Is it necessary?

Reply:
I added a comment and extracted this into a helper method, hoping that naming the function `getTblNotExistError` will make it more clear: https://github.com/liquidata-inc/dolt/pull/299/files#diff-7bbaaa4ef160f9fffed44cef87371fa5R98

This function is supposed to error out before RemoveTables.